src/main.c

```diff
@@ -233,27 +403,120 @@ static void write_spc(FILE *f) {
+		memcpy(new_spc + 0x100 + dstInstruments, &spc_copy[inst_base], inst_size);
+
+		// Copy sample data and pointers...
+		// Remap sample data to new locations to repack them
+		struct {
+			WORD src, len, dst;
+		} sampleMap [NUM_INSTRUMENTS];
+		WORD offset = 0;
+		for (unsigned int i = 0; i < NUM_INSTRUMENTS && sample_pointers[i].start < 0xFF00 && sample_pointers[i].start > 0xFF; i++)
+		{
+			const WORD src_addr = sample_pointers[i].start;
+			WORD len = sample_pointers[i].loop - sample_pointers[i].start;
+			WORD dst_addr = 0;
+
+			// Find if we've already stored this sample somewhere
+			for (unsigned int j = 0; j < i; j++) {
+				if (sampleMap[j].src == src_addr && sampleMap[j].len == len) {
+					dst_addr = sampleMap[j].dst;
+					len = sampleMap[j].len;
+					break;
+				}
+			}
+
+			// If we haven't, add it
+			if (!dst_addr) {
+				unsigned int num_brr_blocks = count_brr_blocks(spc_copy, src_addr);
+				if (num_brr_blocks == 0) {
+					printf("Invalid BRR block. Instrument: %d BRR source: $%x\n", i, src_addr);
+				} else if (offset < 0xFFFF - dstSamples - num_brr_blocks * BRR_BLOCK_SIZE) {
+					dst_addr = dstSamples + offset;
+					printf("Packing sample #%d: $%x size: 0x%x\n", i, dst_addr, num_brr_blocks * BRR_BLOCK_SIZE);
+					memcpy(new_spc + 0x100 + dst_addr, &spc_copy[src_addr], num_brr_blocks * BRR_BLOCK_SIZE);
+
+					offset += num_brr_blocks * BRR_BLOCK_SIZE;
+				} else {
+					printf("Sample #%d is too big. (Destination: $%x, size: 0x%x) Skipped...\n", i, dstSamples + offset, num_brr_blocks * BRR_BLOCK_SIZE);
+				}
+			}
+
+			sampleMap[i].src = src_addr;
+			sampleMap[i].dst = dst_addr;
+			sampleMap[i].len = len;
+			WORD loop_addr = dst_addr + len;
+			printf("Packing sample pointer %d to $%x: %x %x\n", i, dstSamplePointers +0x4*i, dst_addr, loop_addr);
+			memcpy(new_spc + 0x100 + dstSamplePointers + 0x4*i, &dst_addr, 2);
+			memcpy(new_spc + 0x100 + dstSamplePointers + 0x4*i + 0x2, &loop_addr, 2);
 		}
 
-		// Set pattern repeat location. (Not absolutely necessary...)
-		const WORD repeat_address = cur_song.address + 0x2*cur_song.repeat_pos;
-		memcpy(new_spc + 0x140, &repeat_address, 2);
+		// We're done copying the important stuff, now we just need to adjust some pointers in the music program.
+		{
+			// Set pattern repeat location
+			const WORD repeat_address = dstMusic + 0x2*cur_song.repeat_pos;
+			memcpy(new_spc + 0x140, &repeat_address, 2);
 
-		// Set BGM to load
-		const BYTE bgm = selected_bgm + 1;
-		memcpy(new_spc + 0x1F4, &bgm, 1);
+			// Set BGM to load
+			const BYTE bgm = selected_bgm + 1;
+			memcpy(new_spc + 0x1F4, &bgm, 1);
 
-		// Update song address of current BGM within the music program
-		memcpy(new_spc + 0x2F48 + 0x2*bgm, &cur_song.address, 2);
+			// Update song address of current BGM within the music program
+			memcpy(new_spc + 0x2F48 + 0x2*bgm, &dstMusic, 2);
+
+			// Update instrument loader addresses
+			memset(new_spc + 0xA72, dstInstruments & 0xFF, 1);
```

Comment:
That's kinda verbose for writing one byte, isn't it? Was there a reason for that?

```suggestion
			new_spc[0xA72] = dstInstruments & 0xFF;
```

Reply:
It was just to kind of be consistent with the memcpy calls. I've updated them to use assignment instead.